src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1945,8 +1945,14 @@ public void Dispose()
 
                 lock (_buildManager._syncLock)
                 {
-                    _buildManager._projectCacheService?.Result.ShutDown().GetAwaiter().GetResult();
-                    _buildManager._projectCacheService = null;
+                    try
+                    {
+                        _buildManager._projectCacheService?.Result.ShutDown().GetAwaiter().GetResult();
```

Comment:
Nit: Is it guaranteed that it will not get in race condition with line `var projectCacheShutdown = _projectCacheService?.Result.ShutDown();` ?
I believe having protection against double execution inside `Shutdown()` method would be beneficial.

Reply:
I don't think it will race because EndBuild [waits for all submissions to finish](https://github.com/dotnet/msbuild/blob/6819f7ab06c3f43e83ff4059d417395e0af06c01/src/Build/BackEnd/BuildManager/BuildManager.cs#L776-L783) before disposing the plugin. This means it waits for any graph build submissions too, and those will shutdown the cache and set it to null before exiting.

---
In reply to: [583598558](https://github.com/dotnet/msbuild/pull/6110#discussion_r583598558) [](ancestors = 583598558)